DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1

```diff
@@ -117,69 +123,157 @@ function Get-WuaAuSettings
-    }
-    catch [System.Runtime.InteropServices.COMException]
+    $lastHResult = 0
+    $errorCount = 0
+    $running = $true
+
+    while ($running)
     {
-        switch($_.Exception.HResult)
+        try
         {
-            # 0x8024001e    -2145124322    WU_E_SERVICE_STOP    Operation did not complete because the service or system was being shut down.    wuerror.h
-            -2145124322 {
-                Write-Warning 'Got an error that WU service is stopping.  Handling the error.'
-                return $ExceptionReturnValue
-            }
-            # 0x8024402c    -2145107924    WU_E_PT_WINHTTP_NAME_NOT_RESOLVED    Same as ERROR_WINHTTP_NAME_NOT_RESOLVED - the proxy server or target server name cannot be resolved.    wuerror.h
-            -2145107924 {
-                # TODO: add retry for this error
-                Write-Warning 'Got an error that WU could not resolve the name of the update service.  Handling the error.'
-                return $ExceptionReturnValue
-            }
-            # 0x8024401c    -2145107940    WU_E_PT_HTTP_STATUS_REQUEST_TIMEOUT    Same as HTTP status 408 - the server timed out waiting for the request.    wuerror.h
-            -2145107940 {
-                # TODO: add retry for this error
-                Write-Warning 'Got an error a request timed out (http status 408 or equivalent) when WU was communicating with the update service.  Handling the error.'
-                return $ExceptionReturnValue
-            }
-            # 0x8024402f    -2145107921    WU_E_PT_ECP_SUCCEEDED_WITH_ERRORS    External cab file processing completed with some errors.    wuerror.h
-            -2145107921 {
-                # No retry needed
-                Write-Warning 'Got an error that CAB processing completed with some errors.'
-                return $ExceptionReturnValue
+            return Invoke-Command -ScriptBlock $tryBlock -NoNewScope -ArgumentList $argumentList
+            $running = $false
+        }
+        catch [System.Runtime.InteropServices.COMException]
+        {
+            if ($_.Exception.HResult -ne $lastHResult)
+            {
+                $lastHResult = $_.Exception.HResult
+                $errorCount = 0
             }
-            default {
-                throw
+            switch ($_.Exception.HResult)
+            {
+                # 0x8024001e    -2145124322    WU_E_SERVICE_STOP    Operation did not complete because the service or system was being shut down.    wuerror.h
+                -2145124322
+                {
+                    Write-Warning 'Got an error that WU service is stopping.  Handling the error.'
+                    return $ExceptionReturnValue
+                }
+                # 0x8024402c    -2145107924    WU_E_PT_WINHTTP_NAME_NOT_RESOLVED    Same as ERROR_WINHTTP_NAME_NOT_RESOLVED - the proxy server or target server name cannot be resolved.    wuerror.h
+                -2145107924
```

Comment:
Not blocking, but it might be good to refactor this into a function that takes an structure of the errorcode, if it's retryable and warning and iterates over them and does the right thing.

It would reduce code duplication and hopefully make the code easier to read.

Reply:
I've made this change you suggested.